livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py

```diff
@@ -361,32 +376,52 @@ async def send_task():
                 samples_per_channel=samples_50ms,
             )
 
+            has_ended = False
+            last_frame: Optional[rtc.AudioFrame] = None
             async for data in self._input_ch:
-                if isinstance(data, self._FlushSentinel):
+                frames: list[rtc.AudioFrame] = []
+                if isinstance(data, rtc.AudioFrame):
+                    state = self._check_energy_state(data)
+                    if state in (
+                        AudioEnergyFilter.State.START,
+                        AudioEnergyFilter.State.SPEAKING,
+                    ):
+                        if last_frame:
+                            frames.extend(
+                                audio_bstream.write(last_frame.data.tobytes())
+                            )
+                            last_frame = None
+                        frames.extend(audio_bstream.write(data.data.tobytes()))
+                    elif state == AudioEnergyFilter.State.END:
+                        # no need to buffer as we have cooldown period
+                        frames = audio_bstream.flush()
+                        has_ended = True
+                    elif state == AudioEnergyFilter.State.SILENCE:
+                        # buffer the last silence frame, since it could contain beginning of speech
+                        # TODO: improve accuracy by using a ring buffer with longer window
+                        last_frame = data
+                elif isinstance(data, self._FlushSentinel):
                     frames = audio_bstream.flush()
-                else:
-                    frames = audio_bstream.write(data.data.tobytes())
+                    has_ended = True
 
                 for frame in frames:
-                    state = self._audio_energy_filter.update(frame)
-                    if state in (
-                        _AudioEnergyFilter._State.START,
-                        _AudioEnergyFilter._State.SPEAKING,
-                    ):
-                        self._pushed_audio_duration += frame.duration
-                        await ws.send_bytes(frame.data.tobytes())
-                    elif state == _AudioEnergyFilter._State.END:
-                        usage_event = stt.SpeechEvent(
-                            type=stt.SpeechEventType.RECOGNITION_USAGE,
-                            request_id=self._request_id,
-                            alternatives=[],
-                            recognition_usage=stt.RecognitionUsage(
-                                audio_duration=self._pushed_audio_duration
-                            ),
-                        )
-                        self._pushed_audio_duration = 0.0
-                        self._event_ch.send_nowait(usage_event)
+                    self._pushed_audio_duration += frame.duration
+                    await ws.send_bytes(frame.data.tobytes())
+
+                    if has_ended:
```

Comment:
IIUC, has_ended is only true when Flushing when the energy filter is disabled, so the metrics will never be emitted on the VoicePipelineAgent because it is only flushed at shutdown.

I think we should still emit metrics even when energy_filter is False (by always checking for the SpeechEnded from the EnergyFilter avg)

Reply:
it's also set when flush sentinel is seen.

Reply:
I see what you mean, let me see if I can tweak this quickly